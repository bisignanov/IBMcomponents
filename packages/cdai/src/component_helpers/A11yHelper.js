//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

export const keys = {
  SPACE: ' ',
  TAB: 'Tab',
  CLICK: 'click',
  KEYPRESS: 'keypress',
  KEYDOWN: 'keydown',
  KEYUP: 'keyup',
  ENTER: 'Enter',
};

export const handlers = {
  onClick: 'onClick',
  onKeyDown: 'onKeyDown',
  onKeyUp: 'onKeyUp',
  onKeyPress: 'onKeyPress',
};

const defaultValidKeys = [keys.CLICK, keys.SPACE, keys.ENTER];
const defaultHandlers = [handlers.onClick, handlers.onKeyDown];

// triggers a callback if event key matches one in validKeys
export const keyAndClickHandler = (callback, validKeys = defaultValidKeys) => {
  return (evt) => {
    if (evt && (evt.type || evt.key)) {
      if (validKeys.includes(evt.type) || validKeys.includes(evt.key)) {
        callback(evt);
      }
    }
  };
};

export const createInteractionHandler = (
  callback,
  handlers = defaultHandlers,
  validKeys = defaultValidKeys
) => {
  let handler = {};

  for (let i = 0; i < handlers.length; i++) {
    handler[handlers[i]] = keyAndClickHandler(callback, validKeys);
  }
  return handler;
};
