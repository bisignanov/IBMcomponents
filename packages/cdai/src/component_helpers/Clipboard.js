//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

export const copyToClipboard = (event, content) => {
  const field = document.createElement('textarea');
  event.currentTarget.parentNode.appendChild(field);

  field.textContent = content;
  field.select();
  document.execCommand('copy');

  event.currentTarget.parentNode.removeChild(field);
};

export const copyContentToClipboard = (content) => {
  return (event) => copyToClipboard(event, content);
};
