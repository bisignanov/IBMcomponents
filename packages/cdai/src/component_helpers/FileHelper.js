//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

export function getExtensionFromFilename(filename) {
  // invalid input
  if (!filename || typeof filename !== 'string') {
    return;
  }

  const indexOfDot = filename.lastIndexOf('.');

  // no extension
  if (indexOfDot === -1) {
    return;
  }

  return filename.substr(indexOfDot + 1);
}

export function hasValidFilename(filename) {
  var rg1 = /^[^\\/:*?"<>|]+$/; // forbidden characters \ / : * ? " < > |
  var rg2 = /^\./; // cannot start with dot (.)
  var rg3 = /^(nul|prn|con|lpt[0-9]|com[0-9])(\.|$)/i; // forbidden file names
  return rg1.test(filename) && !rg2.test(filename) && !rg3.test(filename);
}

export function hasValidExtension(ext, validExtensions) {
  if (validExtensions.includes('*') || validExtensions.includes(ext)) {
    return true;
  }
  return false;
}

export function parseFileFromUrl(url) {
  if (typeof url === 'string' && url.length > 0) {
    const lastSlashIndex = url.lastIndexOf('/');
    return url.substr(lastSlashIndex + 1);
  }
  return '';
}
