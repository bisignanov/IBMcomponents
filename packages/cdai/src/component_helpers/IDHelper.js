//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

const attribute = 'data-ui-id';

export function idAttributeSelector(identifier) {
  let attr = `[${attribute}="${identifier}"]`;
  return attr;
}

export function fuzzyAttributeSelector(identifier) {
  let attr = `[${attribute}*="${identifier}"]`;
  return attr;
}

export function idAttribute(identifier) {
  let attr = { [attribute]: identifier };
  return attr;
}
