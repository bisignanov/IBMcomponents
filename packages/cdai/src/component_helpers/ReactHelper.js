//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import React from 'react';
import { idePrefix } from '../globals/js/settings';
//This function must be capable of rendering a graphic in multiple different formats:
// 1. if we have a render function, assume react component
// 2. if already a valid react element, just render it
// 3. If this was an object with imgSrc, imgAlt and class name we need to build an img component from them.
export const renderGraphic = (graphicStruct) => {
  // if we have a render function, assume react component
  if (graphicStruct.render && typeof graphicStruct.render === 'function') {
    return React.createElement(graphicStruct);
    // Otherwise, if already a valid react element, just render it
  } else if (React.isValidElement(graphicStruct)) {
    return graphicStruct;
  } else {
    return (
      <img
        src={graphicStruct.imgSrc}
        alt={graphicStruct.imgAlt}
        className={`${idePrefix}--home-image ${
          graphicStruct.imgClassName ? graphicStruct.imgClassName : ''
        }`}
      />
    );
  }
};
