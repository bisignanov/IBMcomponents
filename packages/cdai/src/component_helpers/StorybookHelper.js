//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import { getPathForComponent } from '../../../core/story-structure';

export const getComponentLabel = (componentName) => {
  const title =
    // if the component isn't in the master structure, put it in a lost+found section
    getPathForComponent('a', componentName) ||
    `Cloud & Cognitive/Lost + found/${componentName}`;

  return title;
};
