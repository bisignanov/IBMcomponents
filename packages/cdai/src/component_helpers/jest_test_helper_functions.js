//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React from 'react';
import { shallow, mount } from 'enzyme';

export const getJSXForComponent = (
  component,
  defaultProps = {},
  customProps = {}
) => {
  return React.createElement(component, { ...defaultProps, ...customProps });
};

export const mountComponent = (jsx, returnShallow = true) => {
  let component, componentInstance;

  if (returnShallow) {
    component = shallow(jsx);
  } else {
    component = mount(jsx);
  }

  componentInstance = component.instance();

  let unmount = () => {
    unmountComponent(component);
  };

  return {
    component,
    componentInstance,
    unmount,
  };
};

export const unmountComponent = (component) => {
  component && component.length > 0 && component.unmount();
};
