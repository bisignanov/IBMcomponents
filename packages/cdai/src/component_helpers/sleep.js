export default function sleep(timeInSeconds) {
  return new Promise((resolve) => {
    setTimeout(resolve, timeInSeconds * 1000);
  });
}
