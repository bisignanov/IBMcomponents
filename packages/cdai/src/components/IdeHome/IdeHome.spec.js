//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//

import { IdeHome } from '.';
import {
  withCustomSetTabFunc,
  withCustomSetSectionFunc,
  withNoTasksOrRenderFunc,
  withCollapsibleAbility,
  tasksValidImage,
  withOneSetToRender,
} from './test_assets/testProps.js';
import { idAttributeSelector } from '../../component_helpers/IDHelper';
import * as jth from '../../component_helpers/jest_test_helper_functions.js';

describe('IdeHome unit tests', () => {
  let component, unmount;

  const mountTestComponent = (defaultProps = {}, props = {}) => {
    let test = jth.mountComponent(
      jth.getJSXForComponent(IdeHome, defaultProps, props)
    );
    component = test.component;
    unmount = test.unmount;
    return test;
  };

  afterEach(() => {
    // if unmount defined, call it to clear up any mounted component
    unmount && unmount();
  });

  it('IdeHome tollerates no props being provided, and renders with defaults', () => {
    const testClassNames = {
      'IdeHome-Header-Text': [
        'ide-home-heading-container',
        'bx--col-sm-1',
        'bx--col-md-2',
        'bx--col-lg-6',
        'bx--col-xlg-6',
        'bx--col-max-6',
      ],
      'IdeHome-Header-Image': [
        'ide-home-image-container',
        'bx--col-sm-3',
        'bx--col-md-6',
        'bx--col-lg-10',
        'bx--col-xlg-10',
        'bx--col-max-10',
      ],
    };
    mountTestComponent();
    expect(component.getElements()).toMatchSnapshot();
    for (let [id, classNames] of Object.entries(testClassNames)) {
      classNames.map((className) => {
        expect(
          component.find(idAttributeSelector(id)).hasClass(className)
        ).toEqual(true);
      });
    }
  });

  it('IdeHome renders user custom content when render function is provided', () => {
    const props = withCustomSetTabFunc();
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('tab1')));
    expect(component.exists(idAttributeSelector('tab2')));
  });

  it('IdeHome renders task image when valid react element is provided', () => {
    const props = tasksValidImage();
    mountTestComponent(null, props);
    expect(component.exists('ide-home-tile-img'));
  });

  it('IdeHome renders default sets when render function is not provided', () => {
    const props = withNoTasksOrRenderFunc();
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('tab1')));
    expect(component.exists(idAttributeSelector('tab2')));
  });

  it('IdeHome renders sets as sections when multipleSetType prop is `section`', () => {
    const props = withCustomSetSectionFunc();
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('section1')));
    expect(component.exists(idAttributeSelector('section2')));
  });

  it('IdeHome renders just 1 set if 1 provided', () => {
    const props = withOneSetToRender();
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('tab1')));
  });

  it('IdeHome renders collapsible version of the banner', () => {
    const testClassNames = {
      'IdeHome-Header-Text': [
        'ide-home-heading-container',
        'bx--col-sm-2',
        'bx--col-md-3',
        'bx--col-lg-6',
        'bx--col-xlg-6',
        'bx--col-max-6',
      ],
      'IdeHome-Header-Image': [
        'ide-home-image-container',
        'bx--col-sm-2',
        'bx--col-md-3',
        'bx--col-lg-8',
        'bx--col-xlg-8',
        'bx--col-max-8',
      ],
      'IdeHome-header-toggle-container': [
        'ide-home-header-toggle-container',
        'bx--col-sm-1',
        'bx--col-md-2',
        'bx--col-lg-2',
        'bx--col-xlg-2',
        'bx--col-max-2',
      ],
    };
    const props = withCollapsibleAbility();
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('IdeHome-header-toggle')));
    for (let [id, classNames] of Object.entries(testClassNames)) {
      classNames.map((className) => {
        expect(
          component.find(idAttributeSelector(id)).hasClass(className)
        ).toEqual(true);
      });
    }
  });

  it('renders the collapsed banner when the toggle is pressed', () => {
    const testClassNames = {
      'IdeHome-Header-Text': [
        'ide-home-heading-container',
        'bx--col-sm-3',
        'bx--col-md-6',
        'bx--col-lg-14',
        'bx--col-xlg-14',
        'bx--col-max-14',
      ],
      'IdeHome-header-toggle-container': [
        'ide-home-header-toggle-container',
        'bx--col-sm-1',
        'bx--col-md-2',
        'bx--col-lg-2',
        'bx--col-xlg-2',
        'bx--col-max-2',
      ],
    };
    let props = withCollapsibleAbility();
    props.headerCollapsed = true;
    mountTestComponent(null, props);
    expect(
      component.find(idAttributeSelector('IdeHome-Header')).prop('className')
    ).toEqual('collapsed ide-home-header bx--row');
    for (let [id, classNames] of Object.entries(testClassNames)) {
      classNames.map((className) => {
        expect(
          component.find(idAttributeSelector(id)).hasClass(className)
        ).toEqual(true);
      });
    }
  });

  it('collapses the banner when the toggle is pressed and calls the callback', () => {
    const props = withCollapsibleAbility();
    const mockHeaderHandleCollapseFunc = jest.fn();
    props.headerHandleCollapseFunc = mockHeaderHandleCollapseFunc;
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('IdeHome-header-toggle')));
    component
      .find(idAttributeSelector('IdeHome-header-toggle'))
      .simulate('click');
    expect(mockHeaderHandleCollapseFunc).toHaveBeenCalled();
  });

  it('collapses the banner and calls the callback twice when the toggle is pressed twice', () => {
    const props = withCollapsibleAbility();
    const mockHeaderHandleCollapseFunc = jest.fn();
    props.headerHandleCollapseFunc = mockHeaderHandleCollapseFunc;
    mountTestComponent(null, props);
    expect(component.exists(idAttributeSelector('IdeHome-header-toggle')));
    component
      .find(idAttributeSelector('IdeHome-header-toggle'))
      .simulate('click');
    expect(mockHeaderHandleCollapseFunc).toHaveBeenCalled();
    component
      .find(idAttributeSelector('IdeHome-header-toggle'))
      .simulate('click');
    expect(mockHeaderHandleCollapseFunc).toHaveBeenCalled();
  });
});
