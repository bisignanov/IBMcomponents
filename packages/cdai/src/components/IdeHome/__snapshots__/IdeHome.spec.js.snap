// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeHome unit tests IdeHome tollerates no props being provided, and renders with defaults 1`] = `
[
  <section
    className="ide-home undefined"
    data-ui-id="IdeHome"
  >
    <IdePageContent
      className=""
    >
      <div
        className=" ide-home-header bx--row"
        data-ui-id="IdeHome-Header"
      >
        <div
          className="
                ide-home-heading-container
                bx--col-sm-1
                bx--col-md-2
                bx--col-lg-6
                bx--col-xlg-6
                bx--col-max-6
              "
          data-ui-id="IdeHome-Header-Text"
        >
          <div>
            <p
              className="ide-home-product-heading"
            >
              Product text
            </p>
            <p
              className="ide-home-call-to-action"
            >
              Call to action
            </p>
          </div>
        </div>
        <div
          className="
                  ide-home-image-container
                  bx--col-sm-3
                  bx--col-md-6
                  bx--col-lg-10
                  bx--col-xlg-10
                  bx--col-max-10
                "
          data-ui-id="IdeHome-Header-Image"
        >
          <img
            className="ide--home-image "
          />
        </div>
      </div>
      <div
        className="bx--row bx--grid bx--grid--condensed ide-home-table"
        data-ui-id="IdeHome-Sets"
      />
    </IdePageContent>
  </section>,
]
`;
